DT-1110: Migrate away from @mui/styles Part 1

snf2ye opened the pull request:
## Background
We must discontinue use of @mui/styles in order to migrate from React 17 to 18. ([source](https://mui.com/system/styles/api/))

>  @mui/styles was deprecated with the release of MUI Core v5 in late 2021. It is not compatible with [React.StrictMode](https://react.dev/reference/react/StrictMode) or React 18+, and it will not be updated. Please use [@mui/system](https://mui.com/system/getting-started/) instead.

There is guide for this migration here: https://mui.com/material-ui/migration/migrating-from-jss

Options considered:
- [`sx` prop](https://mui.com/system/getting-started/the-sx-prop/)  + styled components from `@mui/matertial/styles`  -  From migration guide - "We recommend sx API over styled for creating responsive styles or overriding minor CSS". We can use styled components for components with a lot of styles (4+) and for styling that is reused. 
- tss-react - Easiest change, but still would require using copilot per component. It would be more of a lateral step.

This PR uses the first option and uses the `sx` prop alongside styled components. Given that we have to touch every file manually and I think we can utilize copilot for the bulk of the work, I would prefer to do the harder change rather than the lateral change. 

## Steps for Upgrades
**Pre-checks:**
- Review the styles listed at the beginning of the file. Are they all referenced in the component? (Apparently if there are styles that don't match up to a component, it will try to infer where the component with those styles should go and add it) 

1) Run copilot command, specifically in VSCode with Claude 3.5 Sonnet.

>Can you migrate this component away from using JSS and references to '@mui/styles'? We want to instead use a combination of inline styling with the "sx" prop on Material UI components and styled components from @mui/material/styles.

>Let me explain something about how our code is set up: In the existing code, all of the styles currently live in a "styles" object. It contains a list of classes each with their own styles.

>Consider the following example:
```
const styles = (theme: CustomTheme) =>
  createStyles({
    pageRoot: {
      padding: '16px 24px',
      display: 'flex',
      flexDirection: 'column',
      height: '100%',
    },
    jadeTableSpacer: {
      paddingBottom: theme.spacing(12),
    }
});
```

>pageRoot and jadeTableSpacer are classes that we reference in the component. pageRoot has 4 styles and jadeTableSpacer has 1 style. We don't want to use inline styling with the sx for classes that have more than 3 styles, so we want to handle these two classes in different manners. Since pageRoot has 4 styles, we want to use a styled component. Since jadeTableSpacer only has 1 style, we want to use the inline styling with an "sx" prop. We can evaluate other style classes with this same rule: if there are 3 or fewer styles, use the "sx" prop. If there are more than 3 styles, then convert to a styled component.

>When using the "sx" prop, then we must use the material UI version of the component (e.g. Box instead of div and add an import statement at the beginning of the file). We must also switch any css references of direct numbers to include px (e.g. '10px' instead of 10)

2) Apply prettier
3) Review UI and code generated
4) Compare UI/styling between local UI and dev


### Common Issues
- It doesn't handle theme changes well - https://github.com/DataBiosphere/jade-data-repo-ui/commit/30ca7c2d2a6c15a6f8f8c6690781714a04c17cf2

## What step is the copilot upgrade taking? (example output from copilot)
I'll help migrate the SnapshotAccessRequestTable from JSS to use styled components and sx props. 

Let's analyze the styles:
1) More than 3 styles:
-  overlaySpinner (6 styles)
-  openButton (7 styles)
2) Convert to sx props:
- jadeLink (1 style with theme mixin)

Key changes made:
- Removed withStyles HOC and styles object
- Added styled import from @mui/material/styles
- Created styled components for complex styles: OverlaySpinner for loading overlay, RequestButton for the request name button, Converted jadeLink to use sx prop with theme mixin
- Removed classes prop and ClassNameMap interface
- Added Box component for styled span
- Simplified component exports
- Maintained all functionality while modernizing the styling approach

The component maintains the same styling and behavior while using modern MUI patterns.


**Example**
**Before with @mui/styles**
```
const styles = (theme: CustomTheme) => ({
  wrapper: {
    display: 'flex',
    justifyContent: 'center',
    marginTop: '1em',
  }
);
....
 <div className={classes.wrapper}>
...
</div>
```
**With sx prop**
```
 <Box sx={{ display: 'flex', justifyContent: 'center', marginTop: '1em' }}>
...
</Box>
```

##TODO - add example for styled component

## TODO - add more notes about nuances of  the changes from mui



## Components Impacted in this PR
**Updated screenshots as of 01-11-2024**
1) HomeView and 2) DatasetView
<img width="1352" alt="image" src="https://github.com/user-attachments/assets/a386d4ab-d373-4f5b-83dc-f3c4573bc52b" />


3) JobView
<img width="1341" alt="image" src="https://github.com/user-attachments/assets/40948704-9895-49ab-8934-81862af0270a" />


4) SnapshotView
<img width="1345" alt="image" src="https://github.com/user-attachments/assets/c40410c0-fb1d-48bc-96df-b8c3b3a85bb8" />



5) DatasetOverview and 6) DatasetOverviewPanel
<img width="1342" alt="image" src="https://github.com/user-attachments/assets/290978e1-30f4-4da4-977f-7b9797f9b2d7" />
<img width="1342" alt="image" src="https://github.com/user-attachments/assets/1dc37842-404c-4a50-9015-59496228584d" />


7) DatasetAccess and 8) UserList
<img width="1342" alt="image" src="https://github.com/user-attachments/assets/c2554df4-f415-476d-ade8-a99f6b79e51d" />

9) SnapshotAccessRequestTable
<img width="1339" alt="image" src="https://github.com/user-attachments/assets/ca49f54f-929c-44b3-902a-efea57b96290" />

snf2ye committed:
HomeView - converted with copilot

snf2ye committed:
DatasetView - upgrade by copilot

Command used:
Can you migrate this component away from using JSS and references to '@mui/styles' and instead use '@mui/system'? Please use the "sx" prop wherever possible instead of "styled". If you need to use the "sx" prop, then we must use the material UI version of the component (e.g. Box instead of div and add an import statement at the beginning of the file). We must also switch any css references of direct numbers to include px (e.g. '10px' instead of 10)

snf2ye committed:
DatasetView - Copilot decided to go rouge and just add a title??

snf2ye committed:
DatasetView: Copilot dropped the mixins css

Im not sure where it got that maxwidth from

Update references

snf2ye committed:
SnapshotView: Prep for copilot changes

- title isn't used anywhere

add comma

snf2ye committed:
Snapshot View: Update by copilot

Command:
Can you migrate this component away from using JSS and references to '@mui/styles' and instead use '@mui/system'? Please use the "sx" prop wherever possible instead of "styled". If you need to use the "sx" prop, then we must use the material UI version of the component (e.g. Box instead of div and add an import statement at the beginning of the file). We must also switch any css references of direct numbers to include px (e.g. '10px' instead of 10)

snf2ye committed:
SnapshotView: copilot dropped the mixins css

snf2ye committed:
JobView - prep for copilot - remove unused styles

snf2ye committed:
JobView - upgrade by copilot

snf2ye committed:
JobView: cleanup

- Switch again to use theme mixins
- userEmail is not used

snf2ye committed:
Index

I think we need to use @mui/material/styles instead?

This is what is referenced in docs: https://mui.com/material-ui/customization/theming/

remove @mui/styles references from theme

snf2ye committed:
SnapshotAccessRequestTable

snf2ye committed:
DatasetOverview: Remove unused styles

snf2ye committed:
DatasetOverview - Copilot update

snf2ye committed:
DatasetOverviewPanel - update with copilot

snf2ye committed:
DatasetOverviewPanel.jsx - set the Drawer to temporary so it doesn't stick out on the page

snf2ye commented on src/components/JobView.tsx:
```diff
@@ -1,86 +1,64 @@
+interface IProps {
   jobs: Array<JobModel>;
-  dispatch: Dispatch<Action>;
+  dispatch: Dispatch<any>;
```
TODO: Why is this now any instead of Action?

snf2ye commented on src/components/dataset/overview/DatasetOverviewPanel.jsx:
```diff
@@ -294,26 +241,45 @@ function DatasetOverviewPanel(props) {
-            [classes.drawerClose]: !isHelpVisible,
-          }),
+        className={isHelpVisible ? 'drawerOpen' : 'drawerClose'}
+        sx={{
+          top: '112px',
+          bottom: '0px',
+          flexShrink: 0,
+          height: 'initial',
+          zIndex: 10,
+          minWidth: 300,
+          width: '40%',
+          position: 'absolute',
+          transition: () =>
+            theme.transitions.create('width', {
+              easing: theme.transitions.easing.sharp,
+              duration: theme.transitions.duration.enteringScreen,
+            }),
+          '&.drawerClose': {
+            transition: () =>
+              theme.transitions.create('width', {
+                easing: theme.transitions.easing.sharp,
+                duration: theme.transitions.duration.leavingScreen,
+              }),
+            overflowX: 'hidden',
+            width: '0px',
+            minWidth: '0px',
+            [theme.breakpoints.up('sm')]: {
+              width: '0px',
+            },
```
For more complicated styles like this, it may make sense to pull it into a styled component to help with code readability. But, this styling is still only used in one spot so I'm a torn over which is better (leave as sx prop vs. pull into styled component)

fboulnois replied:
Since the refactoring is being done in stages, we can always pull this out into a separate style in the future.

snf2ye committed:
DatasetOverviewPanel.jsx - clean up references

snf2ye committed:
DatasetAccess - remove styles, not referenced in component

snf2ye committed:
UserList - remove unused styles

snf2ye committed:
UserList - update by copilot

snf2ye committed:
JobView - switch back to more specific type

fboulnois commented on src/components/UserList.tsx:
```diff
@@ -1,75 +1,53 @@
+          fontSize: '14px',
+          lineHeight: '22px',
+          fontWeight: '600',
+          color: 'primary.main',
```
Do these need to `import theme from 'modules/theme';` and use `theme.palette.primary.main` instead like in `src/components/DatasetView.tsx`? If so, a number of components in this PR have this issue with `color`.

snf2ye replied:
Great question - I believe it is correct as is - Here's[ the documentation](https://mui.com/system/getting-started/the-sx-prop/?srsltid=AfmBOor9PjYY4oizLrQBEB2KSea8tGUdEU5VACrtpu6Js6RSEbVKr837#basic-example) and the basic example has something very similar. We have ["ThemeProvider" set up in the index](https://github.com/DataBiosphere/jade-data-repo-ui/blob/829daf5eb149d9896b9903ce9cf4c297e8090104/src/index.jsx#L162), so the theme should be available to all components in the repo. 

I'm possibly not referencing the theme correctly in DatasetView, but I wasn't quite sure how to do the expansion of the custom mixin.

fboulnois replied:
Does that also mean that it isn't necessary for `DatasetView.tsx`?

fboulnois commented on src/components/UserList.tsx:
```diff
@@ -1,75 +1,53 @@
+          <Typography
+            sx={{
+              fontStyle: 'italic',
+              color: 'error.contrastText',
```
The original component had `colorPrimary` as well, should this be captured?

snf2ye replied:
Oh I missed that, thanks for catching it. 
It appears the the "(None)" text is correctly picking up the contrast text color, so I don't think we should need both.

fboulnois reviewed (approved):
One main question, otherwise lgtm:

s-rubenstein commented on src/components/DatasetView.tsx:
```diff
@@ -40,56 +23,51 @@ interface IProps extends WithStyles<typeof styles> {
-  },
-);
+  return (
+    <Box sx={{ display: 'flex', justifyContent: 'center', marginTop: '1em' }}>
```
If you want to reduce the diff here, and keep the object abstractions, this could be `<Box sx={styles(theme).wrapper}>` just as easily, and that way we keep our object abstractions while moving to the `sx` property?

To do this you would need to have `styles` be a raw object rather than a method (or call it once if we have a theme we could pass), but that should be okay

snf2ye replied:
That's a really interesting idea, and I'm torn about it. 
I think I'm less interested in making the diff smaller (copilot is doing the heavy lifting here to move things around), and want to think about how we want this to look long term. We often use existing code as a model for new code, so what do we want our best practices to be going forward? 
I think I generally like the sx prop styles to be inline for a small set of changes, but we may want to consider using styled components more frequently to keep a pseudo object abstraction. I could actually build that into the copilot request - like "if there are more than 3 css styles, use a styled component instead of the sx prop".

s-rubenstein replied:
That's my ultimate preference as well. 3 css styles is fine for inline, more gets bulky. I also think if we have a often reused width or something - with lots of constants reused, that I'd abstract.

snf2ye replied:
Turns out that its tricky to get copilot to handle these two cases 🫠 It's not great at counting, apparently (It would catch some components with >3 styles, but not all). 
We may have to do this in two steps (convert to sx prop or styled and then move the rest). I think I'm also having better luck with a newer gpt model in vs code (as opposed to intellij).

rushtong reviewed (approved):
👍🏽

rjohanek commented on src/components/DatasetView.tsx:
```diff
@@ -1,34 +1,17 @@
 import { DatasetRoles } from '../constants';
 
-const styles = (theme: CustomTheme) => ({
-  wrapper: {
```
so just a general comment, this is moving us away from having separate style objects and towards inline styling. Is that the principle that we want to follow? I thought that separate objects were preferred because they are reusable and easier to maintain. But maybe that isn't the case. What are the guidelines from the UI working group on this?

snf2ye replied:
My goal in this PR is to use best practices for Material UI. We're moving from an old Material UI concept (@mui/styles), to a new one - sx prop + styled components. I don't think Terra UI uses Material UI, but I'm glad to still run this by them. 

RE: separate objects that are reusable/easier to maintain: We should use styled components for this case. The "sx prop" should be used for styles that are used in one spot, styled components are reusable and then we use the global theme to share styling variables. I think this is a helpful guide here: https://mui.com/material-ui/customization/how-to-customize

snf2ye committed:
HomeView

snf2ye committed:
UserList - updated by claude copilot

snf2ye committed:
DatasetView

snf2ye committed:
SnapshotView

snf2ye committed:
Job View

snf2ye committed:
DatasetOverview

snf2ye committed:
DatasetOverviewPanel

snf2ye committed:
Remove styles

snf2ye committed:
SnapshotAccessRequestTable.tsx

snf2ye committed:
Use styled from @mui/system, move mixins to be styled components

snf2ye committed:
Fix bug in DatasetOverviewPanel.jsx

Line was still showing up in middle of the page

snf2ye committed:
needed component=span

snf2ye committed:
Getting typescript to recognize that we're using our extended CustomTheme type was quite a pain

- I feel pretty confident that we want to use @mui/material/styles since this is how we define our custom theme

rjohanek commented on src/components/dataset/overview/DatasetOverviewPanel.jsx:
```diff
@@ -231,14 +196,26 @@ function DatasetOverviewPanel(props) {
+                    fontSize: '14px',
+                    lineHeight: '22px',
+                    fontWeight: '600',
+                    color: 'primary.main',
```
just curious, why is this not a styled component? Is something like this just a symptom of copilot not counting very well?

snf2ye replied:
Yeah, it really seems like copilot can't compare numbers all that well 😭  

Here's copilot's notes from this update:

> I'll help migrate the DatasetOverviewPanel from JSS to use styled components and sx props. Let's analyze the styles:
> 
> 1. More than 3 styles:
> - drawer (5 styles)
> - drawerOpen (4 styles with transitions)
> - drawerClose (5 styles with transitions and breakpoints)
> 
> 2. Convert to sx props:
> - root (1 style)
> - helpOverlayCloseButton (1 style)
> - helpText (1 style)
> - drawerPosition (1 style)
> - **header (4 styles)**
> - expandIcon (1 style)
> - noRelationships (3 styles)
> - divider (2 styles)
> - fullViewButton (1 style)
> - fullViewText (1 style)
>
> Key changes made:
> - Removed withStyles HOC and styles object
> - Added styled import from @mui/material/styles
> - Created StyledDrawer component for complex drawer styling
> - Converted all simple styles to sx props
> - Added Box component as root container
> - Removed classes prop and references
> - Simplified component exports
> - Used theme spacing units where appropriate
> - Maintained all functionality while modernizing the styling approach
> - The component maintains the same styling and behavior while using modern MUI patterns.

.

It clearly notes that the "header" class has 4 styles and notes that the first group is for components with more than 3 styles, but still groups it with the sx props. So, I really don't understand why it groups it with the sx props group. But, it still does catch the more complex styles and generally seems to move the components with lots of styles to styled components. I think as long as it is loosely following this rule, then it is okay.

rjohanek commented on src/components/DatasetView.tsx:
```diff
@@ -1,34 +1,18 @@
-    paddingBottom: theme.spacing(8),
-  },
-});
+const Container = styled('div')(({ theme }: { theme: CustomTheme }) => theme.mixins.containerWidth);
```
for variables like container that are used in multiple files do we want to define them in one location and import them?

snf2ye replied:
Excellent point - I think we should use this as an opportunity for identifying components that could be shared. 

I made this a shared component and then realized that "containerWidth" is just setting width to 100%. That seems unnecessary to be a shared theme, so I removed these styled components and the theme property and instead am setting an sx prop here.

rjohanek reviewed (approved):
I think this looks good overall, just a couple non-blocking questions

s-rubenstein commented on src/components/HomeView.tsx:
```diff
@@ -18,76 +15,42 @@ import {
-      fontSize: '1.5rem',
-    },
-  });
+const RootContainer = styled(Box)({
```
I really like this approach - very clean.

s-rubenstein commented on src/components/UserList.tsx:
```diff
@@ -1,75 +1,48 @@
-          [classes.canManageUsers]: canManageUsers,
-        })}
-      >
+      <AccordionDetails data-cy="user-email" sx={canManageUsers ? { pt: 0 } : undefined}>
```
Is `pt: 0` the same as not rendering? Should that be paddingTop?

snf2ye replied:
Yeah, I'll switch back to paddingTop

s-rubenstein commented on src/components/UserList.tsx:
```diff
@@ -1,75 +1,48 @@
+            sx={{
+              fontStyle: 'italic',
+              color: 'error.contrastText',
+              pb: '8px',
```
Same with `pb`? Is that shorthand for paddingBottom? If so, can we use the full CSS version?

s-rubenstein commented on src/components/dataset/overview/DatasetOverviewPanel.jsx:
```diff
@@ -259,7 +236,7 @@ function DatasetOverviewPanel(props) {
                       </dl>
                       {i < dataset.schema.relationships.length - 1 && (
-                        <Divider className={classes.divider} />
+                        <Divider sx={{ mt: '14px', mb: '14px' }} />
```
Same with `mt` and `mb` - I'd prefer the longer versions even if that is functional shorthand.

s-rubenstein reviewed (commented):
I am seeing a lot of `p` or `m` as a shorthand for margin/padding - I don't know if that is an affordance that material provides, but it seems less readable to me than the CSS property - I would rather we stick with the CSS property versions? I commented a few examples

snf2ye committed:
Move to common component

snf2ye committed:
On second thought, let's remove containerWidth

All it is doing is setting the width at 100%. Let's move that to an sx prop rather than a shared component/theme. Let's make this clear what styles we're applying when it is just a simple 100% width.

snf2ye commented:
> I am seeing a lot of `p` or `m` as a shorthand for margin/padding - I don't know if that is an affordance that material provides, but it seems less readable to me than the CSS property - I would rather we stick with the CSS property versions? I commented a few examples

@s-rubenstein  Yeah, this is specifically a feature of the sx prop for mui components: https://mui.com/system/getting-started/the-sx-prop/#spacing
I'm glad to switch these back to the CSS property. If we were really all-in on mui styling, it seems possible that these shortcuts could help with readability of bigger sx props. But, css properties are more readable for me.
Copilot was randomly deciding between using the shorthands and the regular css properties, so we'll have to watch out for this happening in future PRs. We can add a note in the copilot prompt.

snf2ye committed:
Move away from shorthand to more verbose css styling

s-rubenstein reviewed (approved):
LGTM - there are a number of places where we have `14px` vs `14` for styles, but tbh I think that's an existing issue and out of scope to unify to one or the other here.

snf2ye commented:
> LGTM - there are a number of places where we have `14px` vs `14` for styles, but tbh I think that's an existing issue and out of scope to unify to one or the other here.

Oh, good catch - That's something I paid attention to in the first round, but forgot about in the second round. It's actually important that we pay attention to this for spacing styles because of how mui does theme. And, unfortunately, it looks different for styled vs. sx prop 🫠 I'll go back through the changes again.

### sx Prop spacing

With the sx prop, if you leave off "px", then it defaults to it theme spacing, which multiples the value by 8px. So, `sx={margin: 2}` is actually equivalent to `sx={margin: '16px'}`
<img width="839" alt="image" src="https://github.com/user-attachments/assets/4b376ca8-7816-456a-9fe5-e72db1c7a1a1" />

### Styled spacing
With styled components, you have to add `theme.spacing(2)` to use their multiplier shorthand.
styled component: 
* `margin: 2` = `margin: 2px`
* `margin: theme.spacing(2)` = `margin: 16px` = `margin: 16`

<img width="830" alt="image" src="https://github.com/user-attachments/assets/c0c9cf18-b26f-4687-b9bd-ead5fec756c8" />

### Other sx prop styles
To make things even more confusing, these rules appear to not be consistent across sx props. Like sizing (so, height, width, etc.) does NOT default to the theme sizing, but grid does. These guides on sx prop and styled components are vital to getting through these changes. 
https://mui.com/system/getting-started/the-sx-prop
https://mui.com/system/styled/

snf2ye committed:
Add px; update one more test reference
